snowflake/cmd/trafficDrop.go

```diff
@@ -0,0 +1,294 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+
+	"antrea.io/theia/snowflake/pkg/udfs"
+	"antrea.io/theia/snowflake/pkg/utils/timestamps"
+)
+
+const (
+	trafficDropFunctionName           = "traffic_drop"
+	defaultTrafficDropFunctionVersion = "v0.1.0"
+	defaultTrafficDropWaitTimeout     = "5m"
+)
+
+func buildTrafficDropUdfQuery(
+	jobType string,
+	start string,
+	end string,
+	startTs string,
+	endTs string,
+	clusterUUID string,
+	databaseName string,
+	functionVersion string,
+) (string, error) {
+	now := time.Now()
+	detectionID := uuid.New().String()
+
+	var queryBuilder strings.Builder
+
+	queryBuilder.WriteString(`WITH filtered_flows AS (
+SELECT
+	sourceIP,
+	sourcePodName,
+	sourcePodNamespace,
+	destinationIP,
+	destinationPodName,
+	destinationPodNamespace,
+	to_date(flowStartSeconds) as flowStartDate,
+	ingressNetworkPolicyRuleAction,
+	egressNetworkPolicyRuleAction,
+	count(*) as flowNumber
+FROM
+	flows
+WHERE
+	ingressNetworkPolicyRuleAction = 2
+	OR
+	ingressNetworkPolicyRuleAction = 3
+	OR
+	egressNetworkPolicyRuleAction = 2
+	OR
+	egressNetworkPolicyRuleAction = 3
+`)
+
+	var startTime string
+	if startTs != "" {
+		startTime = startTs
+	} else if start != "" {
+		var err error
+		startTime, err = timestamps.ParseTimestamp(start, now)
+		if err != nil {
+			return "", err
+		}
+	}
+	if startTime != "" {
+		fmt.Fprintf(&queryBuilder, `AND
+  flowStartSeconds >= '%s'
+`, startTime)
+	}
+
+	var endTime string
+	if endTs != "" {
+		endTime = endTs
+	} else if end != "" {
+		var err error
+		endTime, err = timestamps.ParseTimestamp(end, now)
+		if err != nil {
+			return "", err
+		}
+	}
+	if endTime != "" {
+		fmt.Fprintf(&queryBuilder, `AND
+  flowEndSeconds < '%s'
+`, endTime)
+	}
+
+	if clusterUUID != "" {
+		_, err := uuid.Parse(clusterUUID)
+		if err != nil {
+			return "", err
+		}
+		fmt.Fprintf(&queryBuilder, `AND
+  clusterUUID = '%s'
+`, clusterUUID)
+	} else {
+		logger.Info("No clusterUUID input, all flows will be considered during abnormal traffic drop detection.")
+	}
+
+	queryBuilder.WriteString(`GROUP BY
+	sourceIP,
+	sourcePodName,
+	sourcePodNamespace,
+	destinationIP,
+	destinationPodName,
+	destinationPodNamespace,
+	flowStartDate,
+	ingressNetworkPolicyRuleAction,
+	egressNetworkPolicyRuleAction
+), processed_flows AS (
+SELECT
+CASE WHEN f.ingressNetworkPolicyRuleAction = 2 OR f.ingressNetworkPolicyRuleAction = 3
+	THEN
+	(CASE WHEN f.destinationPodName IS NOT NULL
+		THEN CONCAT(f.destinationPodNamespace, '#', f.destinationPodName)
```

Comment:
nit: do you think `/` would be a better separator (as that's the separator typically used in K8s between namespace and name)?

Reply:
Sounds good, changed.